File: 3336-find-the-number-of-subsequences-with-equal-gcd/3336-find-the-number-of-subsequences-with-equal-gcd.cpp
class Solution {
public:
    #define ll long long

    ll dp[201][201][201];
    ll mod = 1e9+7;

    ll countPairs(ll i, ll ga,ll gb,vector<int> &nums){
        if(i ==  nums.size()){
            if(ga == gb && ga!=0 ) return 1;
            return 0;
        }
        if(dp[i][ga][gb] != -1) return dp[i][ga][gb];       
        ll ans = countPairs(i+1,ga,gb,nums);
        if(ga == 0)
            ans += countPairs(i+1,nums[i],gb,nums);
        else
            ans += countPairs(i+1,__gcd(ga, (ll)nums[i]),gb,nums);       

        ans%=mod;        

        if(gb == 0)
            ans += countPairs(i+1,ga,nums[i],nums);        

        else
            ans += countPairs(i+1,ga,__gcd(gb,(ll) nums[i]),nums);              

        return dp[i][ga][gb] = ans%mod;

    }

    int subsequencePairCount(vector<int>& nums) {
        memset(dp,-1,sizeof(dp));
        return countPairs(0,0,0,nums);

    }
};

File: 3336-find-the-number-of-subsequences-with-equal-gcd/3336-find-the-number-of-subsequences-with-equal-gcd.py
class Solution:
    mod = 10**9 + 7
    def subsequencePairCount(self, nums: List[int]) -> int:
        n = len(nums)

        @lru_cache(None)
        def countPairs(i, ga, gb):
            if i == n:
                return 1 if ga == gb and ga != 0 else 0

            ans = countPairs(i + 1, ga, gb)

            if ga == 0:
                ans += countPairs(i + 1, nums[i], gb)
            else:
                ans += countPairs(i + 1, gcd(ga, nums[i]), gb)

            ans %= self.mod

            if gb == 0:
                ans += countPairs(i + 1, ga, nums[i])
            else:
                ans += countPairs(i + 1, ga, gcd(gb, nums[i]))

            return ans % self.mod

        return countPairs(0, 0, 0)